### PR TITLE
[WIP] Optionally avoid recordStream calls in ProcessGroupNCCL

### PR DESCRIPTION
**Pull request by mcarilli**
In big-NLP models with various types of model parallelism, some of our internal researchers (@azhurkevich, Eric Harper, Murat Guney), and our own @eqy, often encounter performance-killing allocator thrashing (repeated cudaMalloc/Free calls, even when the process as a whole doesn't go OOM).

Playing with `max_split_size_mb` and `garbage_collection_threshold` didn't help (imo the `garbage_collection_threshold` code is not useful anyway, it tries to extract a free lunch where there's none to be had).

What DID help was commenting out all the recordStream calls in ProcessGroupNCCL. The thrashing behavior disappeared in every repro, presumably because recordStream prevents the allocator from repurposing allocations passed to collectives until those collectives finish from the host's perspective (which could take a while in the ideal scenario where CPU enqueueing runs far ahead of GPU execution).

Of course, recordStreams were there for a reason: to prevent memory in user-facing stream(s) from being repurposed by the allocator while it's still in use by nccl side stream(s). Scenario 0 below sketches the potential problem. Scenario 1 sketches ProcessGroupNCCL's current (recordStream-based) guardrails.
```python
# Scenario 0, unsafe
# "later ops" may use a's memory, and race with "use a..."
a = torch.rand(..., device="cuda")
s.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s):
    use a...
del a
later ops...

# Scenario 1, safe
# a.record_stream(s) ensures the allocator won't repurpose a for "later ops..."
# until any kernels that could have used a in s finish from the host's perspective.
a = torch.rand(..., device="cuda")
s.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s):
    a.record_stream(s)
    use a...
del a
later ops...
```

This PR adds alternative stream-safety guardrails that don't rely on recordStream calls. Sketch:
```python
# Scenario 2, safe
# We never call a.record_stream(s), but we don't release a for repurposing in
# its creation stream until after we sync that creation stream with "use a...",
# so there's no way "later ops..." can act on a's memory in a way that races with "use a..."
a = torch.rand(..., device="cuda")
s.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s):
    use a...
torch.cuda.current_stream().wait_stream(s)
del a
later ops...
```
I tried to port scenario 2 to each collective by holding live references to participating tensors until I'm sure the user-facing stream has been synced with the nccl side stream (ie, until im sure the user has wait()ed on the collective).

Because we're juggling razor blades here and it's hard to test, recordStream avoidance is off by default, and existing default (aka recordStream-based) behavior is unchanged. recordStream avoidance is enabled by setting envvar `NCCL_AVOID_RECORD_STREAMS=1`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

**Comment by mcarilli** [comment]
How should I document this? It doesn't look like there's a dedicated area in https://pytorch.org/docs/stable/distributed.html for variables that govern Pytorch's NCCL bindings.
https://pytorch.org/docs/stable/distributed.html#common-environment-variables mentions NCCL's own variables.
`NCCL_BLOCKING_WAIT` and `NCCL_ASYNC_ERROR_HANDLING` are described in the docstring for [`init_process_group()`'s `timeout`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group) (but clearly `NCCL_AVOID_RECORD_STREAMS=1` doesn't belong there).
Some variables, cf
```
mcarilli:c10d$ grep -ir parseEnv
ProcessGroupNCCL.cpp:  blockingWait_ = parseEnvVarFlag(NCCL_BLOCKING_WAIT);
ProcessGroupNCCL.cpp:  asyncErrorHandling_ = parseEnvVarFlag(NCCL_ASYNC_ERROR_HANDLING);
ProcessGroupNCCL.cpp:  desyncDebug_ = parseEnvVarFlag(NCCL_DESYNC_DEBUG);
ProcessGroupNCCL.cpp:  avoidRecordStreams_ = parseEnvVarFlag(NCCL_AVOID_RECORD_STREAMS);
ProcessGroupNCCL.cpp:  if (parseEnvVarFlag(ENABLE_NCCL_HEALTH_CHECK)) {
```
don't appear to be documented in https://pytorch.org/docs/stable/distributed.html at all.

What would you like me to do?

**Comment by mcarilli** [comment]
Alex Z and @eqy confirmed the PR eliminates thrashing on their workloads without apparent corruption (on workloads where we expect bitwise reproducibility, loss curves over a few iterations and without the PR are bitwise accurate). In other words, we think it's good to go once https://github.com/pytorch/pytorch/pull/76861#issuecomment-1145109206 is resolved.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -551,6 +557,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   blockingWait_ = parseEnvVarFlag(NCCL_BLOCKING_WAIT);
   asyncErrorHandling_ = parseEnvVarFlag(NCCL_ASYNC_ERROR_HANDLING);
   desyncDebug_ = parseEnvVarFlag(NCCL_DESYNC_DEBUG);
+  avoidRecordStreams_ = parseEnvVarFlag(NCCL_AVOID_RECORD_STREAMS);
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:560]: Nit: can we prefix env's with `TORCH_DISTRIBUTED_` instead of `NCCL_`? People sometimes think these are NCCL env's, so we are starting to avoid that.
Also, do we plan to maintain this env for the longer term?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1537,6 +1554,17 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
...
+  // avoidRecordStreams_ note:
+  // send, recv, and irecv should be ok with avoidRecordStreams,
+  // However, for isend, I don't think the API requires the user
+  // to wait() on the returned handle, so ProcessGroupNCCL can't know
+  // when it's safe to release the input back to the allocator,
+  // and the present call has no way to know it's not an isend.
+  // Therefore, we warn and fall back to the typical recordStream logic:
+  TORCH_WARN_ONCE(!avoidRecordStreams_,
+      "NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point "
+      "collectives.");
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1566]: Nit: "point-to-point collectives" --> "point-to-point communications"

**Review by rohan-varma — COMMENTED** [review]
Had some discussion with @mrshenli, there is an initial high level question we're wondering about. A lot of this is quoted from @mrshenli:

It's not fully clear why avoiding `record_stream` would help. In Scenario 1 safe and Scenario 2 safe cases, `a` will be deleted when the last references on `a` got gc-ed. In both scenarios, this happens after using `a` , the only difference is that in (1), CUDACachingAllocator hold's `a`'s reference, whereas in (2), user manages it.

It seems that we can achieve the same result as the safe scenario 2 in scenario 1, without removing `record_stream`. In particular, could we just add `wait_stream(s)` before call to `del a`, which should also avoid allocator thrashing.

**Comment by eqy** [comment]
Would the difference be in how quickly `a`'s allocation can be repurposed? It sounds like in scenario 1, _all_ enqueued work in the collective must be finished, whereas in the PR's implementation, the references are only held for the duration of the collective.

If I'm understanding correctly, it sounds like adding a `wait_stream` would require changes in a user-provided training loop, and with user-knowledge of the allocation and deallocation pattern alongside NCCL usage. While this could work in theory, it's perhaps involved to modify a training loop in e.g., `Megatron-LM` and then for other comparable training setups that could run into this issue. The main advantage of this PR is that it would allow a workaround for use cases with intricate patterns of allocation an deallocation (such as those with pipeline + model parallelism) without detailed investigation and knowledge.

**Comment by eqy** [comment]
CC @mrshenli 
Are there any remaining blockers for this PR? Our current perspective is that as it is an opt-in mechanism and benefits existing known/tested use-cases, it would be safe and beneficial to merge.

**Comment by mrshenli** [comment]
@kwen2501 is our communication expert. Hey Ke, do you mind help review and stamp this PR if it looks OK to you? Thanks!

**Review by kwen2501 — COMMENTED** [review]
Hi, I am coming back to share some general thoughts:
1. it is great to see that the allocator thrashing is gone;
2. though the change uses an optional environment variable and does not change the default behavior, it introduces a second path throughout the code. For example, the `if` conditions and notes are scattered at quite a few places. This can potentially make the code more difficult to maintain or for new comers to develop, and easier to create bugs.
3. so the tipping point would depend on 1 vs 2. How general do you observe the thrashing issue? (with respect to application spectrum)
4. you probably won't want to give your customer an env var in production. So for the longer term, if we know one approach is better than the other, we should switch to that one, instead of keeping both in parallel.
